main.cpp

```diff
@@ -6319,673 +6330,698 @@ string cpu_reg(int val) {
-        std::copy(sm.begin(), sm.end(), std::back_inserter(tmp));
-        insts.push_back(tmp);
-        contents = sm.suffix();
-    }
-
-    for (vector<string> sm : insts) {
-        uint32_t val;
-        if (sm.size() < 7) {
-            fos << sm[0] << " was too small\n";
-            continue;
-        }
-        if (sm[3].length()) {
-            // Clang
-            int b0, b1, b2, b3 = 0;
-            get_int("0x" + sm[5], b0);
-            get_int("0x" + sm[6], b1);
-            get_int("0x" + sm[2], b2);
-            get_int("0x" + sm[4], b3);
-            val = b0 + (b1 << 8) + (b2 << 16) + (b3 << 24);
-        } else {
-            // GCC
-            int b0, b1 = 0;
-            get_int("0x" + sm[5] + sm[6], b0);
-            get_int("0x" + sm[2] + sm[4], b1);
-            val = b0 + (b1 << 16);
-        }
-
-        uint32_t mcrbits = val & 0xff100010;
-        bool mcr = false;
-        uint32_t mccrbits = (val & 0xfff00000) >> 20;
-        bool mccr = false;
-        uint32_t cdpbits = val & 0xff000010;
-        bool cdp = false;
-        string inst;
-
-        switch (mcrbits) {
-            case 0xee000010:
-                mcr = true;
-                inst = "mcr";
-                break;
-            case 0xfe000010:
-                mcr = true;
-                inst = "mcr2";
-                break;
-            case 0xee100010:
-                mcr = true;
-                inst = "mrc";
-                break;
-            case 0xfe100010:
-                mcr = true;
-                inst = "mrc2";
-                break;
-            default:
-                break;
+static bool parse_hex(const std::string& value, uint32_t& out) {
+    try {
+        size_t pos = 0;
+        out = std::stoull(value, &pos, 16);
+        if (pos != value.length()) {
+            return false;
         }
+        return true;
+    } catch (std::invalid_argument& e) {
+        return false;
+    } catch (std::out_of_range& e) {
+        return false;
+    }
+}
 
-        switch (mccrbits) {
-            case 0xec4:
-                mccr = true;
-                inst = "mcrr";
-                break;
-            case 0xfc4:
-                mccr = true;
-                inst = "mcrr2";
-                break;
-            case 0xec5:
-                mccr = true;
-                inst = "mrrc";
-                break;
-            case 0xfc5:
-                mccr = true;
-                inst = "mrrc2";
-                break;
-            default:
-                break;
-        }
+bool coprodis_command::decode_line(uint32_t val, char *buf, size_t buf_len) {
+    uint32_t mcrbits = val & 0xff100010;
+    enum {
+        NONE = 0,
+        MCR = 0x01,
+        MCR2 = 0x02,
+        MRC = 0x04,
+        MRC2 = 0x08,
+        MCRR = 0x10,
+        MCRR2 = 0x20,
+        MRRC = 0x40,
+        MRRC2 = 0x80,
+        CDP = 0x100,
+    } type = NONE;
+    uint32_t mccrbits = (val & 0xfff00000) >> 20;
+    uint32_t cdpbits = val & 0xff000010;
```

Comment:
Hmm... `mcrbits = val & 0xff100010` and `cdpbits = val & 0xff000010` so it looks like `cdpbits` is actually a **subset** of `mcrbits`?

If the intention is that `val` only ever decodes to _one_ of the `type` enums, perhaps there ought to be a `if (type != NONE) {` around the `switch (mccrbits) {` and `switch (cdpbits) {` blocks?

(And I guess you could left-shift the `case` values in the `switch (mccrbits)` by 20 bits, and then you would't need the `>> 20` when masking off `mccrbits`?)